Refine effects based on optimizer-derived information

Keno opened the pull request:
The optimizer may be able to derive information that is not available to inference. For example, it may SROA a mutable value to derive additional constant information. Additionally, some effects, like :consistent are path-dependent and should ideally be scanned once all optimizations are done. Now, there is a bit of a complication that we have generally so far taken the position that the optimizer may do non-IPO-safe optimizations, although in practice we never actually implemented any. This was a sensible choice, because we weren't really doing anything with the post-optimized IR other than feeding it into codegen anyway. However, with irinterp and this change, there's now two consumers of IPO-safely optimized IR. I do still think we may at some point want to run passes that allow IPO-unsafe optimizations, but we can always add them at the end of the pipeline.

With these changes, the effect analysis is a lot more precise. For example, we can now derive :consistent for these functions:
```
function f1(b)
    if Base.inferencebarrier(b)
        error()
    end
    return b
end

function f3(x)
    @fastmath sqrt(x)
    return x
end
```
and we can derive `:nothrow` for this function:
```
function f2()
    if Ref(false)[]
        error()
    end
    return true
end
```

Keno committed:
Refine effects based on optimizer-derived information

The optimizer may be able to derive information that is not available
to inference. For example, it may SROA a mutable value to derive
additional constant information. Additionally, some effects, like
:consistent are path-dependent and should ideally be scanned once
all optimizations are done. Now, there is a bit of a complication
that we have generally so far taken the position that the optimizer
may do non-IPO-safe optimizations, although in practice we never
actually implemented any. This was a sensible choice, because
we weren't really doing anything with the post-optimized IR other
than feeding it into codegen anyway. However, with irinterp and
this change, there's now two consumers of IPO-safely optimized
IR. I do still think we may at some point want to run passes
that allow IPO-unsafe optimizations, but we can always add them
at the end of the pipeline.

With these changes, the effect analysis is a lot more precise.
For example, we can now derive :consistent for these functions:
```
function f1(b)
    if Base.inferencebarrier(b)
        error()
    end
    return b
end

function f3(x)
    @fastmath sqrt(x)
    return x
end
```
and we can derive `:nothrow` for this function:
```
function f2()
    if Ref(false)[]
        error()
    end
    return true
end
```

vtjnash reviewed (dismissed):
> Now, there is a bit of a complication that we have generally so far taken the position that the optimizer may do non-IPO-safe optimizations, although in practice we never actually implemented any. This was a sensible choice, because we weren't really doing anything with the post-optimized IR other than feeding it into codegen anyway. However, with irinterp and this change, there's now two consumers of IPO-safely optimized IR. I do still think we may at some point want to run passes that allow IPO-unsafe optimizations, but we can always add them at the end of the pipeline.

We don't do as many as we should, but there is a long way between doing only a few and being prohibited from doing any ever. For a direct example, `Base.inferencebarrier` is required to be optimized away, but must not improve IPO information afterwards. Other similar examples may include 1. the legality of removing boundscheck depends on having determined whether it is guaranteed to be not inlined (which premise is broken by this PR). 2. folding Expr(:new) with undef fields to a constant and other similar operations with undefined bits such as accessing `.parameters` or doing math with floating point numbers assumes IPO cannot see that change to the implementation 3. DCE cannot mark a value as unused after all loads are removed (changing it to readnone), since later IPO effects may incorrectly assume this argument could be optimized away in the caller

Keno commented:
> We don't do as many as we should, but there is a long way between doing only a few and being prohibited from doing any ever. 

You're not prohibited from doing any ever, you're just prohibited from doing them before the analysis pass runs. There's no problem doing more things afterwards.

> For a direct example, `Base.inferencebarrier` is required to be optimized away, but must not improve IPO information afterwards.

Yes, that needed to be adjusted in this PR.

> Other similar examples may include 1. the legality of removing boundscheck depends on having determined whether it is guaranteed to be not inlined (which premise is broken by this PR). 

This PR is on the path of overhauling this mechanism. It's possible there's some extra tweaking that needs to happen here in this intermediate step.

> folding Expr(:new) with undef fields to a constant and other similar operations with undefined bits such as accessing `.parameters` or doing math with floating point numbers assumes IPO cannot see that change to the implementation 

I don't believe we do this in the current pass pipeline.

> DCE cannot mark a value as unused after all loads are removed (changing it to readnone), since later IPO effects may incorrectly assume this argument could be optimized away in the caller

We don't currently derive such information interprocedurally.

vtjnash commented:
> you're just prohibited from doing them before the analysis pass runs

In particular though, this means the analysis pass must run before inlining, since inlining already ran all passes. Which seems hardly indistinguishable from not being able to do any useful optimizations?

> This PR is on the path of overhauling this mechanism

I am talking about the default environment, which is not affected by that PR. In the default cases where it is not forced on or forced off, the effect of inlining on boundscheck branches is known to not be consistent, but that inconsistency is usually removed by the inlining pass. Though I suppose that falls into the larger bucket above that inlining is not allowed before the refinement analysis, so it might not matter as an independent point.

Keno commented:
As discussed today, we'll stop doing :boundscheck transforms in inlining and move those to codegen instead. Then we can cache the IR as IPO safe. At whatever point in the future we want to do IPO unsafe IR transforms at the julia level, we'll have a callback from codegen back into julia that does that.

Keno committed:
Don't delete code that may be needed by irinterp

I was seeing a fair number of inference instabilities on master
and tracked this down to us deleting the inferred code after
codegen making it unavailable to irinterp.

Keno committed:
Correctly taint :nothrow effect on GotoIfNot with non-Bool argument

We were usually doing this correctly, but not in the corner case where
both branches ended up in the same place.

Keno committed:
Fixup

Keno committed:
Fixup unstable

Keno committed:
Respect termination consistent-cy

vtjnash commented:
How did we decide to deal with the fact this means that the result of codegen (e.g. validity of using concrete eval) will no longer be consistent with the IPO flags on the same CodeInfo object?

Keno commented:
> How did we decide to deal with the fact this means that the result of codegen (e.g. validity of using concrete eval) will no longer be consistent with the IPO flags on the same CodeInfo object?

You mean CodeInstance? They already have two sets of flags, but if something is determined for IPO, it should always be valid even after codegen transforms.

vtjnash commented on base/compiler/optimize.jl:
```diff
@@ -509,7 +711,7 @@ function run_passes(
     @pass "compact 2" ir = compact!(ir)
```
This pass currently seems to not be IPO safe, since the access to `ci.propagate_inbounds` (and `:boundscheck` removal folding) is not currently IPO safe

Keno replied:
Yes, I'm working on that right now.

Keno committed:
Delay :boundscheck resolution until codegen time

Keno committed:
Corrections to UB behavior, get rid of inbounds effects

Also include UB in optimizer-refinement and get rid of the inbounds
effect, replacing it by a conditional ub effect, which does much
the same thing, but with more precise, path-dependent semantics.

Keno committed:
Emit compilerbarrier as identity in codegen

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -2529,6 +2494,20 @@ function abstract_eval_statement_expr(interp::AbstractInterpreter, e::Expr, vtyp
+            else
+                t = Union{}
+            end
+        elseif !hasintersect(condt, Bool)
```
```suggestion
        elseif !hasintersect(widenconst(condt), Bool)
```

aviatesk commented on base/compiler/effects.jl:
```diff
@@ -139,10 +136,13 @@ const EFFECT_FREE_IF_INACCESSIBLEMEMONLY = 0x01 << 1
+
+const EFFECTS_TOTAL    = Effects(ALWAYS_TRUE,  ALWAYS_TRUE,  true,  true,  true,  ALWAYS_TRUE,  ALWAYS_TRUE,  true)
+const EFFECTS_THROWS   = Effects(ALWAYS_TRUE,  ALWAYS_TRUE,  false, true,  true,  ALWAYS_TRUE,  ALWAYS_TRUE,  true)
+const EFFECTS_UNKNOWN  = Effects(ALWAYS_FALSE, ALWAYS_FALSE, false, false, false, ALWAYS_FALSE, ALWAYS_FALSE, true) # unknown mostly, but it's not overlayed and noinbounds at least (e.g. it's not a call)
```
```suggestion
const EFFECTS_UNKNOWN  = Effects(ALWAYS_FALSE, ALWAYS_FALSE, false, false, false, ALWAYS_FALSE, ALWAYS_FALSE, true) # unknown mostly, but it's not overlayed at least (e.g. it's not a call)
```

aviatesk commented on base/compiler/effects.jl:
```diff
@@ -193,18 +190,18 @@ is_nothrow(effects::Effects)             = effects.nothrow
 is_notaskstate(effects::Effects)         = effects.notaskstate
 is_inaccessiblememonly(effects::Effects) = effects.inaccessiblememonly === ALWAYS_TRUE
-is_noub(effects::Effects)                = effects.noub
+is_noub(effects::Effects, noinbounds=true) = effects.noub === ALWAYS_TRUE || (noinbounds && effects.noub === NOUB_IF_NOINBOUNDS)
```
```suggestion
is_noub(effects::Effects, noinbounds::Bool=true) = effects.noub === ALWAYS_TRUE || (noinbounds && effects.noub === NOUB_IF_NOINBOUNDS)
```

aviatesk commented on base/compiler/effects.jl:
```diff
@@ -193,18 +190,18 @@ is_nothrow(effects::Effects)             = effects.nothrow
 
 # implies `is_notaskstate` & `is_inaccessiblememonly`, but not explicitly checked here
-is_foldable(effects::Effects) =
+is_foldable(effects::Effects, noinbounds=true) =
```
```suggestion
is_foldable(effects::Effects, noinbounds::Bool=true) =
```

aviatesk commented on base/compiler/effects.jl:
```diff
@@ -193,18 +190,18 @@ is_nothrow(effects::Effects)             = effects.nothrow
 
-is_foldable_nothrow(effects::Effects) =
-    is_foldable(effects) &&
+is_foldable_nothrow(effects::Effects, noinbounds=true) =
```
```suggestion
is_foldable_nothrow(effects::Effects, noinbounds::Bool=true) =
```

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -15,27 +15,29 @@ const SLOT_USEDUNDEF    = 32 # slot has uses that might raise UndefVarError
 # NOTE make sure to sync the flag definitions below with julia.h and `jl_code_info_set_ir` in method.c
 
-const IR_FLAG_NULL        = 0x00
+const IR_FLAG_NULL        = UInt32(1)
```
```suggestion
const IR_FLAG_NULL        = UInt32(0)
```

aviatesk commented on base/reflection.jl:
```diff
@@ -1647,9 +1647,9 @@ function infer_effects(@nospecialize(f), @nospecialize(types=default_tt(f));
         match = match::Core.MethodMatch
         frame = Core.Compiler.typeinf_frame(interp,
-            match.method, match.spec_types, match.sparams, #=run_optimizer=#false)
+            match.method, match.spec_types, match.sparams, #=run_optimizer=#true)
```
New optional argument `post_optimization_analysis::Bool=true` might be useful for testing/debugging purposes. We can add it later when needed though.

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -853,14 +847,8 @@ function concrete_eval_eligible(interp::AbstractInterpreter,
-    end
     mi = result.edge
-    if mi !== nothing && is_foldable(effects)
+    if mi !== nothing && is_foldable(effects, !stmt_taints_inbounds_consistency(sv))
```
Maybe we need a similar change to `concrete_eval_invoke`?

Keno replied:
Actually, I don't think it's required here at all. Both concrete eval and semi concrete eval are safe here independent of the state of the inbounds flag.

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+
+    cfg = nothing
+    domtree = nothing
+    function get_augmented_domtree()
```
Probably better to set up a new struct 
```julia
struct AugumentedDomtree
    cfg::CFG
    domtree::DomTree
end
```
and make `get_augumented_domtree` return the struct instead for better inferrability.

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+        boundscheck = stmt.args[end]
+        argextype(boundscheck, ir) === Bool || return false
+        isa(boundscheck, SSAValue) || return false
+        return true
```
Include `isexpr(ir[boundscheck][:stmt], :boundscheck)` here?

Keno replied:
This is meant to check for having a boundscheck argument rather than the boundscheck expr, I'll rename the function.

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+                else isa(stmt, GotoIfNot)
+                    bb = block_for_inst(ir, idx)
+                    sa, sb = ir.cfg.blocks[bb].succs
+                    for succ in iterated_dominance_frontier(ir.cfg, BlockLiveness((sa, sb), nothing), get!(lazydomtree))
```
```suggestion
                    for succ in iterated_dominance_frontier(ir.cfg, BlockLiveness(ir.cfg.blocks[bb].succs, nothing), get!(lazydomtree))
```
and then we can keep typing `BlockLiveness.def_bbs::Vector{Int}`.

aviatesk commented on base/compiler/ssair/ir.jl:
```diff
@@ -1390,7 +1410,7 @@ function process_node!(compact::IncrementalCompact, result_idx::Int, inst::Instr
             # Rename phi node edges
             let bb_rename_pred=bb_rename_pred
-                map!(i::Int32->bb_rename_pred[i], stmt.edges, stmt.edges)
+                map!(i::Int32->i == 0 ? 0 : bb_rename_pred[i], stmt.edges, stmt.edges)
```
Is this change necessary for this PR, or a fix for some related issues?

Keno replied:
It fixed some test failure, but I don't think it's related to this PR other than by coincidence. There's a special `i == 0` predecessor for entry into catch blocks.

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+        end
+    end
+
+    function scan_inconsistency!(inst::Instruction, idx)
```
```suggestion
    function scan_inconsistency!(inst::Instruction, idx::Int)
```

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+        return stmt_inconsistent
+    end
+
+    function scan_stmt!(inst, idx, lstmt, bb)
```
```suggestion
    function scan_stmt!(inst::Instruction, idx::Int, lstmt::Int, bb::Int)
```

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+        if isexpr(stmt, :enter)
+            # try/catch not yet modeled
+            had_trycatch = true
+            return true
```
```suggestion
            return false
```
Shouldn't return `false` to bail out the scan?

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+    if !scan!(scan_stmt!, scanner, true)
+        if !all_retpaths_consistent
+            # No longer any dataflow concerns, just scan the flags
+            scan!(scanner, false) do inst, idx, lstmt, bb
```
```suggestion
            scan!(scanner, false) do inst::Instruction, idx::Int, lstmt::Int, bb::Int
```

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+                return
+            end
+        else
+            callback(idx, stmt)
```
```suggestion
            callback(idx)
```
`stmt::PhiNode` seems to be never used.

aviatesk commented:
> I was seeing a fair number of inference instabilities on master
> and tracked this down to us deleting the inferred code after
> codegen making it unavailable to irinterp.

Doesn't absint-based constprop' usually lead to better inference? I'm in favor of preserving sources for irinterp, but I want to assert my understanding that irinterp is faster but less precise than the absint constprop'.

Keno committed:
Address review and fix rebase mistakes

Keno commented:
> Doesn't absint-based constprop' usually lead to better inference? I'm in favor of preserving sources for irinterp, but I want to assert my understanding that irinterp is faster but less precise than the absint constprop'.

I don't think it's necessarily always true that irinterp is less precise than absint. For example, irinterp can in theory propagate through more complicated sroa patterns. I didn't track down exactly what was going on in this case, but the failing test was arrayops 1746.

Keno committed:
Switch :boundscheck to fully use extra Expr arg

Per in-person review from Jameson.

Keno committed:
Support noub refinement in irinterp

Keno commented:
@nanosoldier `runtests()`

Keno committed:
Update doctest

Keno commented:
MRE for the Plots issue seen on PkgEval:
```
julia> Plots._cbar_unique([nothing], "fill alpha")
Internal error: encountered unexpected error in runtime:
BoundsError(a=Array{Int64, (87,)}[1, 1, 0, 1, 1, 2, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 3, 0, 1, 1, 2, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0], i=(94,))
ijl_bounds_error_ints at /home/keno/julia/src/rtutils.c:201
getindex at ./essentials.jl:13 [inlined]
#346 at ./compiler/ssair/ir.jl:1723 [inlined]
foreachssa at ./compiler/utilities.jl:373
maybe_erase_unused! at ./compiler/ssair/ir.jl:1722
```

Keno commented:
Non-Plots reproducer:
```
@inline function bar()
	x = 1
	try
		x = 2
		if Ref{Bool}(false)[]
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			x = 3
		end
	catch
		return x
	end
end

function foo()
	if Ref{Bool}(false)[]
		bar()
	end
	return 4
end
```

Keno committed:
Fix IncrementalCompact of PhiCNode with cfg mod

Keno commented:
@nanosoldier `runtests(["Sherlock", "NonconvexIpopt", "OptimizationNLopt", "Test", "InferOpt", "NonconvexNOMAD", "OptimizationOptimisers", "GaussianKDEs", "Reel", "StochasticAD", "WaterWaves1D", "ChargeTransport", "NormalizingFlows", "EnKF", "NonconvexJuniper", "BAT", "NeuralQuantumState", "Robotlib", "ReferenceTests", "PerformanceProfilingHttpEndpoints", "OptimalTransmissionRouting", "ClimaCorePlots", "TexasHoldem", "FluxOptTools", "DecisionMakingPolicies", "NonconvexCore", "NonconvexPavito", "NonconvexMMA", "ConditionalDists", "Pseudospectra", "FourierSpaces", "ProtoBuf", "ReinforcementLearningCore", "RemoteFiles", "Diagonalizations", "SurrogatesAbstractGPs", "RRTMGP", "DifferentiableFlatten", "KernelAbstractions", "ProbabilityBoundsAnalysis", "PlantBiophysics", "SurrogatesMOE", "MIRTjim", "Integrals", "UltraDark", "MLJGaussianProcesses", "BaytesMCMC", "Boltz", "OptimizationMOI", "Plots", "RetentionParameterEstimator", "MIRT", "ObjectOriented", "BaytesDiff", "WaveguideQED", "EarthSciMLBase", "Sunny", "RegularizedProblems", "ReactionSensitivity", "NonconvexPercival", "ConScape", "ImplicitDifferentiation", "SBMLToolkit", "OptimizationMetaheuristics", "OptimizationOptimJL", "BM3DDenoise", "ParameterHandling", "ModelWrappers", "RAPIDS", "GeoEstimation", "ReliabilityOptimization", "NonconvexNLopt", "BigRiverQTLPlots", "Anatta", "MultiscaleGraphSignalTransforms", "GeoStatsPlots", "SurrogatesFlux", "ProfileEndpoints", "PtFEM", "LowLevelParticleFilters", "CoDa", "SyntacticModels", "NonconvexUtils", "ADRIA"])`

Keno commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50805#issuecomment-1698415764) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/e9b4026_vs_a84882e/report.md).

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50805#issuecomment-1698415196) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c84f526_vs_8ace9c5/report.html) is available.

Keno commented:
> [Your benchmark job](https://github.com/JuliaLang/julia/pull/50805#issuecomment-1698415764) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/e9b4026_vs_a84882e/report.md).

I looked into this - the couple percent in general are fine. This does more work, so that's not unexpected. I looked into the large regression in `["inference", "abstract interpretation", "many_const_calls"]` - that turns out to just be because the benchmark forcibly disables the optimizer and this PR moves the hack that we used to have in inference to prove consistency of a literal `getfield(x, f, Expr(:boundscheck))` into the dataflow pass instead. Since this is my benchmark, I'm just gonna go ahead and say this is fine. `may_optimize() = false` is not really an optimized configuration (it already messes with precision because it changes the inlining cost model). We should revisit whether it makes sense as a benchmark. 

I'm gonna go ahead and merge this, I have a few more things that can make this better, but I'll save them for later before this PR gets even bigger than it already is.
